src/tiled/command.cpp

```diff
@@ -30,39 +30,69 @@
 using namespace Tiled;
 using namespace Tiled::Internal;
 
+QString Command::finalWorkingDirectory() const
+{
+    QString finalWorkingDirectory = workingDirectory;
+
+    return replaceVariables(finalWorkingDirectory, true);
+}
+
 QString Command::finalCommand() const
 {
-    QString finalCommand = command;
+    QString finalCommand = QString(QLatin1String("%1 %2")).arg(executable).arg(arguments);
+
+    return replaceVariables(finalCommand);
+}
+
+QString Command::replaceVariables(const QString &string, bool flag) const
+{
+    QString finalString = string;
 
     // Perform variable replacement
     if (Document *document = DocumentManager::instance()->currentDocument()) {
         const QString fileName = document->fileName();
 
-        finalCommand.replace(QLatin1String("%mapfile"),
-                             QString(QLatin1String("\"%1\"")).arg(fileName));
+        finalString.replace(QLatin1String("%mapfile"),
+                            QString(QLatin1String("\"%1\"")).arg(fileName));
 
         QFileInfo fileInfo(fileName);
         QString mapPath = fileInfo.absolutePath();
-        finalCommand.replace(
+
+        finalString.replace(
             QLatin1String("%mappath"),
             QString(QLatin1String("\"%1\"")).arg(mapPath));
 
         if (MapDocument *mapDocument = qobject_cast<MapDocument*>(document)) {
             if (const Layer *layer = mapDocument->currentLayer()) {
-                finalCommand.replace(QLatin1String("%layername"),
-                                     QString(QLatin1String("\"%1\"")).arg(layer->name()));
+                finalString.replace(QLatin1String("%layername"),
+                                    QString(QLatin1String("\"%1\"")).arg(layer->name()));
             }
         }
 
         if (MapObject *currentObject = dynamic_cast<MapObject *>(document->currentObject())) {
-            finalCommand.replace(QLatin1String("%objecttype"),
-                                 QString(QLatin1String("\"%1\"")).arg(currentObject->type()));
-            finalCommand.replace(QLatin1String("%objectid"),
-                                 QString(QLatin1String("\"%1\"")).arg(currentObject->id()));
+            finalString.replace(QLatin1String("%objecttype"),
+                                QString(QLatin1String("\"%1\"")).arg(currentObject->type()));
+            finalString.replace(QLatin1String("%objectid"),
+                                QString(QLatin1String("\"%1\"")).arg(currentObject->id()));
+        }
+    }
+
+    if (flag) {
```

Comment:
Ok. Let me try to explain what I was thinking about: I would move the logic for `executablepath` into `finalWorkingDirectory` and then rename `flag` to something like `quoteVariableValues`, which is then false for the working directory case --> quotes are not replaced. Is that more clear?